TESTS/host_tests/pyusb_basic.py

```diff
@@ -734,6 +812,410 @@ def control_stress_test(dev, log):
+            ctrl_error.set()
+
+    delayed_halt = Timer(delay, timer_handler)
+    delayed_halt.start()
+    end_ts = time.time() + 1.5 * delay
+    try:
+        while time.time() < end_ts and not ctrl_error.is_set():
+            loopback_ep_test(ep_out, ep_in, ep_out.wMaxPacketSize)
+    except usb.core.USBError as err:
+        if err.errno not in (32, 110):
+            raise_unconditionally(lineno(), 'Unexpected error ({!r}).'.format(err))
+        if ctrl_error.is_set():
+            raise_unconditionally(lineno(), 'Halting endpoint {0.bEndpointAddress:#04x} failed'
+                                  .format(ep_to_halt))
+    else:
+        raise_unconditionally(lineno(), 'Halting endpoint {0.bEndpointAddress:#04x}'
+                              ' during transmission did not raise USBError.'
+                              .format(ep_to_halt))
+
+
+USB_ERROR_FMT = str('Got {0!r} while testing endpoints '
+                    '{1.bEndpointAddress:#04x}({1.wMaxPacketSize:02}) and '
+                    '{2.bEndpointAddress:#04x}({2.wMaxPacketSize:02}) with '
+                    'a random payload of {3} B.')
+
+
+def ep_test_data_correctness(dev, log, verbose=False):
+    """Test data correctness for every OUT/IN endpoint pair.
+
+    Given a USB device with multiple OUT/IN endpoint pairs
+    When the host sends random payloads up to wMaxPacketSize in size
+        to an OUT endpoint of the device,
+        and then the device sends data back to host using an IN endpoint
+    Then data sent and received by host is equal for every endpoint pair
+    """
+    cfg = dev.get_active_configuration()
+    for intf in cfg:
+        log('interface {}, alt {} -- '.format(intf.bInterfaceNumber, intf.bAlternateSetting), end='')
+        if intf.bAlternateSetting == 0:
+            log('skipping the default AlternateSetting')
+            continue
+        log('running tests')
+        intf.set_altsetting()
+
+        bulk_out, bulk_in = find_ep_pair(intf, usb.ENDPOINT_TYPE_BULK)
+        interrupt_out, interrupt_in = find_ep_pair(intf, usb.ENDPOINT_TYPE_INTERRUPT)
+        iso_out, iso_in = find_ep_pair(intf, usb.ENDPOINT_TYPE_ISOCHRONOUS)
+
+        if verbose:
+            log('\tbulk_out      {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(bulk_out))
+            log('\tbulk_in       {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(bulk_in))
+            log('\tinterrupt_out {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(interrupt_out))
+            log('\tinterrupt_in  {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(interrupt_in))
+            log('\tiso_out       {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(iso_out))
+            log('\tiso_in        {0.bEndpointAddress:#04x}, {0.wMaxPacketSize:02} B'.format(iso_in))
+
+        if verbose:
+            log('Testing OUT/IN data correctness for bulk endpoint pair.')
+        for payload_size in range(1, bulk_out.wMaxPacketSize + 1):
```

Comment:
Zero is a valid size, so it might not hurt to test that value as well. If you were seeing errors with this earlier #6922 should fix the errors.

Reply:
Updated; thanks for the reminder!